hbase-server/src/main/java/org/apache/hadoop/hbase/mapreduce/LoadIncrementalHFilesJob.java

```diff
@@ -0,0 +1,348 @@
+          new LoadIncrementalHFiles.LoadQueueItem(family, hfilePath);
+      queue.add(loadQueueItem);
+    }
+
+    @Override protected void cleanup(Context context) throws IOException, InterruptedException {
+      int queueSize = queue.size();
+      String tableStr = context.getConfiguration().get(TABLE_NAME);
+      try (Connection connection = ConnectionFactory.createConnection(context.getConfiguration());
+          Table table = connection.getTable(TableName.valueOf(tableStr));
+          RegionLocator regionLocator = connection.getRegionLocator(TableName.valueOf(tableStr))) {
+        loadIncrementalHFiles.doBulkloadFromQueue(queue, table, regionLocator, connection);
+      }
+      context.getCounter("Bulkload", "Files Loaded").increment(queueSize);
+    }
+  }
+
+  public static class BulkLoadInputFormat extends InputFormat<Text, Text> {
+
+    private static final Log LOG = LogFactory.getLog(BulkLoadInputFormat.class);
+
+    @Override public List<InputSplit> getSplits(JobContext jobContext)
+        throws IOException, InterruptedException {
+      Configuration configuration = jobContext.getConfiguration();
+
+      List<InputSplit> result = new ArrayList<>();
+      List<BulkoadInputWritable> bulkoadInputWritables = new ArrayList<>();
+
+      int depth = configuration.getInt(DEPTH, 2);
+      if (depth != 2 && depth != 3) {
+        throw new IllegalStateException("Invalid depth " + depth);
+      }
+      Set<String> validFamilies = getValidFamiliesSet(configuration);
+
+      try {
+        Path rootPath = new Path(configuration.get(ROOT_DIR));
+        FileSystem fs = rootPath.getFileSystem(configuration);
+
+        for (FileStatus child : fs.listStatus(rootPath)) {
+          if (!child.isDirectory()) {
+            LOG.info("Skipping non dir " + child.getPath());
+            continue;
+          }
+          if (child.getPath().getName().startsWith("_")) {
+            LOG.info("Skipping " + child.getPath());
+            continue;
+          }
+          if (depth == 2) {
+            bulkoadInputWritables
+                .addAll(createSplitsForFamilyDirectory(fs, child.getPath(), validFamilies));
+          } else {
+            for (FileStatus family : fs.listStatus(child.getPath())) {
+              bulkoadInputWritables
+                  .addAll(createSplitsForFamilyDirectory(fs, family.getPath(), validFamilies));
+            }
+          }
+        }
+
+        int maxMapTasks = configuration.getInt(MAX_MAP_TASK, 1);
+
+        int filesPerMap = bulkoadInputWritables.size() / maxMapTasks;
+        for (List<BulkoadInputWritable> partition : Lists
+            .partition(bulkoadInputWritables, filesPerMap)) {
+          BulkoadInputWritableArray bulkoadInputWritableArray = new BulkoadInputWritableArray();
+          bulkoadInputWritableArray
+              .set(partition.toArray(new BulkoadInputWritable[partition.size()]));
+          result.add(new BulkloadInputSplit(bulkoadInputWritableArray));
+        }
```

Comment:
So we will always default to a single mapper? How can this outperform non-distributed LoadIncrementalHFiles?

Reply:
The idea was to keep the same performance(roughly) as the original version of the bulkload, and allow the user to tweak these settings to obtain better performance.